clippy_lints/src/semicolon_block.rs

```diff
@@ -0,0 +1,131 @@
+use clippy_utils::{get_parent_expr_for_hir, get_parent_node};
+use rustc_errors::Applicability;
+use rustc_hir::{Block, Expr, ExprKind, Node, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// For () returning expressions, check that the semicolon is inside the block.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// For consistency it's best to have the semicolon inside/outside the block. Either way is fine and this lint suggests inside the block.
+    /// Take a look at `semicolon_outside_block` for the other alternative.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # fn f(_: u32) {}
+    /// # let x = 0;
+    /// unsafe { f(x) };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # fn f(_: u32) {}
+    /// # let x = 0;
+    /// unsafe { f(x); }
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub SEMICOLON_INSIDE_BLOCK,
+    restriction,
+    "add a semicolon inside the block"
+}
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// For () returning expressions, check that the semicolon is outside the block.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// For consistency it's best to have the semicolon inside/outside the block. Either way is fine and this lint suggests outside the block.
+    /// Take a look at `semicolon_inside_block` for the other alternative.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # fn f(_: u32) {}
+    /// # let x = 0;
+    /// unsafe { f(x); }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # fn f(_: u32) {}
+    /// # let x = 0;
+    /// unsafe { f(x) };
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub SEMICOLON_OUTSIDE_BLOCK,
+    restriction,
+    "add a semicolon outside the block"
+}
+declare_lint_pass!(SemicolonBlock => [SEMICOLON_INSIDE_BLOCK, SEMICOLON_OUTSIDE_BLOCK]);
+
+impl LateLintPass<'_> for SemicolonBlock {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &Block<'_>) {
+        semicolon_inside_block(cx, block);
+        semicolon_outside_block(cx, block);
+    }
+}
+
+fn semicolon_inside_block(cx: &LateContext<'_>, block: &Block<'_>) {
+    if !block.span.from_expansion()
+    && let Some(tail) = block.expr
+    && let Some(block_expr @ Expr { kind: ExprKind::Block(_, _), ..}) = get_parent_expr_for_hir(cx, block.hir_id)
+    && let Some(Node::Stmt(Stmt { kind: StmtKind::Semi(_), span, .. })) = get_parent_node(cx.tcx, block_expr.hir_id)
+    {
+        let expr_snip = snippet_with_macro_callsite(cx, tail.span, "..");
+
+        let mut suggestion: String = snippet_with_macro_callsite(cx, block.span, "..").to_string();
+
+        if let Some((expr_offset, _)) = suggestion.rmatch_indices(&*expr_snip).next() {
+            suggestion.insert(expr_offset + expr_snip.len(), ';');
+        } else {
+            return;
+        }
```

Comment:
This would create equally large suggestions for large blocks

A multispan suggestion could be a good option here, one span to remove the semicolon and one to add a new one

Something along the lines of `span.with_start(block_expr.span.end())` should get the span of the semicolon

Reply:
Done that, though I am having problems with macro test, I tried using `.source_callsite()` on the spans to no avail, as that always spans the entire `macro_call!();` (including the semi). Not sure how to fix that

Reply:
Hmm, that's difficult. As far as I can see the `.source_callsite()` of the statement doesn't include the semi, but there's nothing I can see that does include it (for `semicolon_outside_block`)

Maybe if the span is from a macro expansion you could use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.mac_call_stmt_semi_span, ignoring macros would also be an option

Reply:
Hm, actually maybe it was the other way around, never containing the semi, will check that tomorrow, thanks!